src/Servers/Kestrel/Core/test/ConnectionDispatcherTests.cs

```diff
@@ -69,5 +70,45 @@ public async Task OnConnectionCompletesTransportPipesAfterReturning()
             mockPipeWriter.Verify(m => m.Complete(It.IsAny<Exception>()), Times.Once());
             mockPipeReader.Verify(m => m.Complete(It.IsAny<Exception>()), Times.Once());
         }
+
+        [Fact]
+        public async Task OnConnectionFiresOnCompleted()
+        {
+            var serviceContext = new TestServiceContext();
+            var dispatcher = new ConnectionDispatcher(serviceContext, _ => Task.CompletedTask);
+
+            var connection = new Mock<TransportConnection> { CallBase = true }.Object;
+            connection.ConnectionClosed = new CancellationToken(canceled: true);
+            var completeFeature = connection.Features.Get<IConnectionCompleteFeature>();
+
+            Assert.NotNull(completeFeature);
+            object stateObject = new object();
+            object callbackState = null;
+            completeFeature.OnCompleted(state => { callbackState = state; return Task.CompletedTask; }, stateObject);
+
+            await dispatcher.OnConnection(connection);
+
+            Assert.Equal(stateObject, callbackState);
+        }
+
+        [Fact]
+        public async Task OnConnectionOnCompletedExceptionCaught()
+        {
+            var serviceContext = new TestServiceContext();
+            var dispatcher = new ConnectionDispatcher(serviceContext, _ => Task.CompletedTask);
+
+            var connection = new Mock<TransportConnection> { CallBase = true }.Object;
+            connection.ConnectionClosed = new CancellationToken(canceled: true);
+            var completeFeature = connection.Features.Get<IConnectionCompleteFeature>();
+
+            Assert.NotNull(completeFeature);
+            object stateObject = new object();
+            object callbackState = null;
+            completeFeature.OnCompleted(state => { callbackState = state; throw new InvalidTimeZoneException(); }, stateObject);
```

Comment:
`InvalidTimeZoneException` interesting choice.

Reply:
It's always fun to find a weird exception that would get your attention if it was ever thrown. My personal favorite is java has a HeadlessException.